Create chapter about rust-lang/rust issue triaging (#716)
rust-lang/rust-forge: src/release/issue-triaging.md

# Issue triaging

This page is about the `rust-lang/rust` repository. Other repositories may have different processes.

Tracking issues (label `C-tracking-issue`) don't fit into this procedure and are treated differently.

## Motivation

The `rust-lang/rust` repository has thousands of issues and hundreds of people working on it.
It is impossible for all people to check and solve issues. The goals of triaging are connecting
issues to the relevant people, and helping them be more effective at fixing the issue.

> In practice, it is unrealistic for all issues to be solved quickly and found the by right people.

Reviewer: ```suggestion
In practice, it is unrealistic for all issues to be solved quickly and found by right people.
```

Through applications of labels we make the issue tracker more searchable for future reference,
so that people in the future have an easier time finding related issues or issues they are interested
in working on.

Triaging can be done by **everyone**, no matter your permissions. We encourage everyone to help here,
as triaging work is highly parallelizable and easy to get started with.

## Initial triaging

When an issue is opened, it gets the `needs-triage` label. This ensures that every issue gets an initial

> look and that no issue is forgotten, or that when it is forgotten, it is at least visibly forgotten by still having the label.

Author: I guess I should replace the "forgotten" with "ignored". This ensures no issues are totally ignored, and labels give them a higher chance of coming back at some point.

`needs-triage` is an initial checkpoint. The effort needed to get an issue past the label should be minimal.

To do the initial triage and remove the `needs-triage` label, the following conditions should be fulfilled/considered.

> It's okay if not all of these are always considered, treat it as guideline, not a hard check list. It is also not exhaustive.

Reviewer 2: ```suggestion
It's okay if not all of these are always considered; treat it as a guideline, not a hard checklist. It is also not exhaustive.
```

> - The issue should make sense, that is it should present a problem.

Reviewer 2: ```suggestion
- The issue should make sense, that is, it should present a problem.
```

>     - For example, if an issue is a question about Rust in general, the issue should be closed and the user redicted to URLO/Discord.

Reviewer 2: ```suggestion
    - For example, if an issue is a question about Rust in general, the issue should be closed and the user redirected to URLO/Discord.
```

      You can of course answer the question too :) (but make sure to mention that the user should go to URLO/Discord next time).
- Add appropriate labels ([Labels](#labels))
    - Specifically, `T-*` and `C-*` are the most relevant
- If the issue contains no reproduction but needs one, ask for one and add the `S-needs-repro` label
- If the issue could benefit from bisecting the regression, add `E-needs-bisection` (or do the bisection yourself)
- Does this issue require nightly? Add `requires-nightly`.
- Is the issue a regression? Apply the `regression-untriaged` label (or figure out what regression it is exactly)
- If you happen to know people who this issue is relevant to, ping them.
    - For example, write `cc @ThatPerson` if `ThatPerson` has been working a lot on the problematic feature recently
- Does this issue require incomplete or internal features? Add `requires-{incomplete,internal}-features`.

> For applying and removing labels, unprivileged users can use rustbot.

Reviewer 2: ```suggestion
For applying and removing labels, unprivileged users can use **@rustbot** to add or remove [the labels allowed by the `triagebot.toml` configuration](https://github.com/rust-lang/rust/blob/master/triagebot.toml).
```

For example, `@rustbot label +T-compiler +C-bug +A-linkage +O-macos -needs-triage`.

To see a list of all labels, check out the "labels" page next to the search bar in the issue tracker.

> ## Further triaging

Reviewer 3: I wonder if we should mention `needs-triage-legacy` here: “If you have time and energy, you can look through NTL issues from time to time and triage them”.

> ## Further triaging

Reviewer 3: (Minor) Could you add a sentence encouraging people to <kbd>needs-triage</kbd> when closing an issue or when the issue has already been closed (or re-open the latter if deemed necessary) since closing an issue is a form of triage?

Although I've noticed that some others do this as well from time to time, I find myself mass-removing <kbd>needs-triage</kbd> from closed issues every so often.

Reviewer 4: Could you elaborate a bit on why this is necessary? What is the use case for including closed issues in needs-triage queries?

Reviewer 3: It's not super important.
(1) Sometimes (on the rarer side) issue authors suddenly close their issue by themselves, e.g., if they find a workaround or if to them it appears the issue has been fixed. In such case, a closed <kbd>needs-triage</kbd> issue might indicate that rustc devs should take another look at it.
(2) More often, <kbd>needs-triage</kbd> issues get fixed by a PR (via *Fixes*) but now the closed issue doesn't have any labels that might be helpful to future people searching the issue tracker for open and/or closed issues which they might do to get the full picture of an area of the compiler. E.g., in my early rustc dev days I used to look for issues similar to one I wanted to fix and see how they were fixed.

Author: I'm not sure whether this is worth it. The best way to do this is either automation (rustbot) or just mass-removing the label on closed issues by hand (which cannot be done by unprivileged users). I don't think anyone should be running around leaving rustbot comments on closed issues to remove needs-triage. I'll just keep this out for now.

For issues that have been through the initial triaging step (that is, don't have the `needs-triage` label anymore), there are usually
still things that can be improved. There are often many more labels that could be applied (using rustbot again if you don't have privileges).

Additionally, old (there is no clear definition of old yet, but something on the order of months) `S-needs-repro` issues can be closed
if there is no way to make progress without a reproduction. This requires privileges, but if you don't have them, you can just link the issue
on Zulip (for example in `t-release/triage` or `general`) and someone with privileges can close it for you.

Another useful thing to do is go through `E-needs-mcve` and `E-needs-bisection` issues and creating minimizations or bisecting the issue
(using [cargo-bisect-rustc](`https://github.com/rust-lang/cargo-bisect-rustc`)). When you provide one, you can also remove the label
using rustbot (`@rustbot label -E-needs-bisection`).

> ## Labels

Reviewer: There is a section in the rustc dev guide that covers this. Should probably consolidate these into a single place with a link from the either.

Author: created #717

There are many different labels that can be applied to issues.

- `needs-triage`: signals that an issue is new and needs initial triage
- `T-*`: Specifies the team or teams that this issue is relevant to, for example compiler, types or libs

> - `C-*`: Specifies the category of the label, for example a bug, tracking issue or discussion

Reviewer 3: Should we mention here that <kbd>A-diagnostic</kbd> issues typically don't / shouldn't be marked with <kbd>C-bug</kbd> but just with <kbd>D-*</kbd> if applicable?

- `O-*`: For platform-specific issues, specifies the platform (architecture or operating system). For example macos, aarch64, windows
- `A-*`: The areas that the issue is relevant to, for example linkage, patterns, diagnostics
- `F-*`: When the issue concerns a specific (usually unstable) feature

> - `requires-nightly`: This issue is not relevant to the stable compiler

Reviewer 3: I wonder if we should expand upon this. For example, personally speaking I only add this label if it's a nightly-to-nightly regression (i.e., a super recent regression) but typically not if I also add <kbd>F-*</kbd> labels since that implies nightly in most cases.

Reviewer 5: I think it's good to add it alongside with `F-*` because you can't really search for "has any `F-*` labels"

Reviewer 6: Should all `A-rustdoc-json`/`A-cranelift` issues also have `requires-nightly`?

Author: I don't know what it should be about exactly. I think `requires-nightly` is most useful in combination with, say `I-unsound`. `I-unsound requires-nightly` makes it clear that hey, this issue is bad, not but affecting stable, so everything is fine for now.

Author: created #718

- `requires-{incomplete,internal}-features`: This issue requires an incomplete or internal feature. The latter often means that the issue
    should be closed in accordance with compiler [MCP 620](https://github.com/rust-lang/compiler-team/issues/620).
- `regression-*`: Labels for tracking issues that are regressions.
- `D-*`: Labels for diagnostics issue.

> - `I-*`: Different labels about the nature (originally, importance) of a bug. For example ICE, slow code, heavy code, crashes, unsoundness.

Reviewer 5: how does nomination thread into this?

Author: It doesn't, this category makes no sense really. It was supposed to be "importance" I guess but like.. no that makes no sense. I guess I should mention it, that it also contains other labels.

- `P-*`: Priority labels. Applied using the [Compiler Prioritization procedure](../compiler/prioritization.md)
- `S-*`: The status of an issue, for example S-needs-repro.

> - `E-*`: Calls for participation, for example to minimize an issue (E-needs-mcve) or because there is mentoring available (E-mentor).

Reviewer 3: Should we call out <kbd>E-needs-test</kbd> here specifically to explain that it's used to marked open issues that have been fixed by some PR already but that we'd like to see a regression test added to close it officially?

> - `E-*`: Calls for participation, for example to minimize an issue (E-needs-mcve) or because there is mentoring available (E-mentor).

Reviewer 3: Should we call out <kbd>E-{easy,medium,hard}</kbd> separately here mentioning that you should only add them if you are an expert in the given area of the compiler etc. and know for certain how difficult it would be for someone else to fix it? Maybe that's obvious though

Reviewer 7: I feel that the compiler experts often evaluate things as easier than they actually are due to the [curse of knowledge](https://en.wikipedia.org/wiki/Curse_of_knowledge). Experts are also not immune to being ambushed by yak-shaving problems: I have seen things be evaluated as _relatively_ simple by an expert, and the issue actually happens to involve at least one more yak than expected. These labels feel inherently dubious to me.

Reviewer 8: I agree; if anything we should label issues as mentored and separately advocate for experts to mentor issues. That has a much better chance of resulting a smooth contributor experience than "this seems easy"

In practice, it is unrealistic for all issues to be solved quickly and found by right people.
Through applications of labels we make the issue tracker more searchable for future reference,
so that people in the future have an easier time finding related issues or issues they are interested
in working on.

Triaging can be done by **everyone**, no matter your permissions. We encourage everyone to help here,
as triaging work is highly parallelizable and easy to get started with.

## Initial triaging

When an issue is opened, it gets the `needs-triage` label. This ensures that every issue gets an initial
look and that no issue is ignored, or that when it is ignored, it is at least visibly ignored by still having the label.

`needs-triage` is an initial checkpoint. The effort needed to get an issue past the label should be minimal.

To do the initial triage and remove the `needs-triage` label, the following conditions should be fulfilled/considered.
It's okay if not all of these are always considered; treat it as a guideline, not a hard checklist. It is also not exhaustive.

- The issue should make sense, that is, it should present a problem.

>     - For example, if an issue is a question about Rust in general, the issue should be closed and the user redirected to URLO/Discord.

Reviewer 9: I'd probably explicitate the links (adding the references)

```suggestion
    - For example, if an issue is a question about Rust in general, the issue should be closed and the user redirected to [URLO] / [Discord] support forums.

[URLO]: https://users.rust-lang.org
[Discord]: https://discord.gg/rust-lang
```

      You can of course answer the question too :) (but make sure to mention that the user should go to URLO/Discord next time).
- Add appropriate labels ([Labels](#labels))
    - Specifically, `T-*` and `C-*` are the most relevant
- If the issue contains no reproduction but needs one, ask for one and add the `S-needs-repro` label
- The issue is the wrong place for some kinds of feature requests. Tell the author about it.
    - Library API requests should follow [its processes](https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html).

>     - Language changes should be redirected to [IRLO](https://internals.rust-lang.org/) or Zulip (t-lang).

Reviewer 9: ```suggestion
    - Language changes should be redirected to [IRLO](https://internals.rust-lang.org/) or Zulip ([t-lang](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang)).
```
